src/pcp/types.rs

```diff
@@ -123,120 +104,96 @@ impl ValueParam for CountValueParam {
 ///
 /// [BBCG+19]: https://ia.cr/2019/188
 #[derive(Clone, Debug, PartialEq, Eq)]
-pub struct Sum<F: FieldElement> {
-    data: Vec<F>,
+pub struct Sum<F> {
+    bits: usize,
     range_checker: Vec<F>,
 }
 
 impl<F: FieldElement> Sum<F> {
-    /// Constructs a new summand. The value of `summand` must be in `[0, 2^bits)`.
-    pub fn new(summand: u64, bits: u32) -> Result<Self, PcpError> {
-        let summand = usize::try_from(summand).unwrap();
-        let bits = usize::try_from(bits).unwrap();
+    /// Return a new [`Sum`] type parameter. Each value of this type is an integer in range `[0,
+    /// 2^bits)`.
+    pub fn new(bits: usize) -> Self {
+        Self {
+            bits,
+            range_checker: poly_range_check(0, 2),
+        }
+    }
+}
 
-        if bits > (size_of::<F::Integer>() << 3) {
-            return Err(PcpError::Value(
+impl<F: FieldElement> Type for Sum<F> {
+    type Measurement = F::Integer;
+    type Field = F;
+
+    fn encode(&self, summand: &F::Integer) -> Result<Vec<F>, PcpError> {
+        // XXX Move this check to constructor
```

Comment:
I meant to resolve these "XXX"s before starting review. I'll resolve them before we merge.